### PR TITLE
PHOENIX-5559 Fix remaining issues with Long viewIndexIds.

### PR DESCRIPTION
**Pull request by lhofhansl**
See https://issues.apache.org/jira/browse/PHOENIX-5559

### CODE REVIEW

File: phoenix-core/src/main/java/org/apache/phoenix/coprocessor/BaseScannerRegionObserver.java [hunk 1]
```diff
@@ -178,8 +178,11 @@ private static void throwIfScanOutOfRegion(Scan scan, Region region) throws DoNo
             byte[] expectedUpperRegionKey =
                     scan.getAttribute(EXPECTED_UPPER_REGION_KEY) == null ? scan.getStopRow() : scan
                             .getAttribute(EXPECTED_UPPER_REGION_KEY);
-            isStaleRegionBoundaries = expectedUpperRegionKey != null &&
-                    Bytes.compareTo(upperExclusiveRegionKey, expectedUpperRegionKey) != 0;
+
```
Comment (lhofhansl) [inline phoenix-core/src/main/java/org/apache/phoenix/coprocessor/BaseScannerRegionObserver.java]: The following is the tricky part. Previous in order to detect a stale region boundary cache we only looked at the upper region bound. That work fine as long as do not have an open scanner that has already reached the upper half of the region when we split. When the scan restarts after the split finishes, the scan restarts in the new upper daugther region and the upper might be the same (it definitely is when this was the last region of the table).

That always was a problem but since we're starting with larger number of the viewIndexId this is just more likely. My solution is to use the Scan's expected start key and compare that. For everything I tested that does the right thing.

File: phoenix-core/src/main/java/org/apache/phoenix/coprocessor/MetaDataEndpointImpl.java [hunk 2]
```diff
@@ -2082,7 +2082,7 @@ private long getViewIndexSequenceValue(PhoenixConnection connection, String tena
         long sequenceTimestamp = HConstants.LATEST_TIMESTAMP;
         try {
             connection.getQueryServices().createSequence(key.getTenantId(), key.getSchemaName(), key.getSequenceName(),
-                    Long.MIN_VALUE, 1, 1, Long.MIN_VALUE, Long.MAX_VALUE, false, sequenceTimestamp);
+                    Short.MIN_VALUE, 1, 1, Long.MIN_VALUE, Long.MAX_VALUE, false, sequenceTimestamp);
```
Comment (lhofhansl) [inline phoenix-core/src/main/java/org/apache/phoenix/coprocessor/MetaDataEndpointImpl.java]: This is the core change. In order to work with both short and long view index ids, we start with Short.MIN_VALUE, which can be encoded both as short and long.

The bulk of the rest are test fixes for tests that happen to verify the generate index id.

File: phoenix-core/src/main/java/org/apache/phoenix/util/MetaDataUtil.java [hunk 3]
```diff
@@ -941,6 +941,16 @@ public static IndexType getIndexType(List<Mutation> tableMetaData, KeyValueBuild
         return null;
     }
 
+    public static PDataType<?> getIndexDataType(List<Mutation> tableMetaData,
+            KeyValueBuilder builder, ImmutableBytesWritable value) {
+        if (getMutationValue(getPutOnlyTableHeaderRow(tableMetaData),
```
Comment (gjacoby126) [inline phoenix-core/src/main/java/org/apache/phoenix/util/MetaDataUtil.java]: nit: comment would be really nice here. I _think_ you're looking up the view index id config setting from syscat, and that makes sense, but the code's a mouthful without explanation :-)

File: phoenix-core/src/main/java/org/apache/phoenix/coprocessor/BaseScannerRegionObserver.java (hunk 1 above)
Comment (lhofhansl) [reply phoenix-core/src/main/java/org/apache/phoenix/coprocessor/BaseScannerRegionObserver.java]: The main part I could use a double-check on... Did I understand the purpose of SCAN_ACTUAL_START_ROW correctly?
From looking at the code and going by the name it's the start row of the scan that we can use in Phoenix to restart the scan (and that can happen exactly when a split occurred and the scanner was restarted by HBase with the new region start key - assuming it fell into the second daughter region). If that is indeed the right interpretation that this should be correct.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/BaseTenantSpecificViewIndexIT.java [hunk 4]
```diff
@@ -148,10 +148,10 @@ private void createAndVerifyIndex(Connection conn, String viewName, String table
                             + "CLIENT MERGE SORT", QueryUtil.getExplainPlan(rs));
         } else {
             String expected = saltBuckets == null ? 
-                    "CLIENT PARALLEL 1-WAY RANGE SCAN OVER _IDX_" + tableName + " [" + Long.toString(Long.MIN_VALUE + expectedIndexIdOffset) + ",'" + tenantId + "','" + valuePrefix + "v2-1']\n"
+                    "CLIENT PARALLEL 1-WAY RANGE SCAN OVER _IDX_" + tableName + " [" + Long.toString(Short.MIN_VALUE + expectedIndexIdOffset) + ",'" + tenantId + "','" + valuePrefix + "v2-1']\n"
```
Comment (ChinmaySKulkarni) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/BaseTenantSpecificViewIndexIT.java]: nit: Use `Short.toString` instead.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/BaseTenantSpecificViewIndexIT.java [hunk 5]
```diff
@@ -148,10 +148,10 @@ private void createAndVerifyIndex(Connection conn, String viewName, String table
...
                             + "    SERVER FILTER BY FIRST KEY ONLY" :
-                    "CLIENT PARALLEL 3-WAY RANGE SCAN OVER _IDX_" + tableName + " [0," + Long.toString(Long.MIN_VALUE + expectedIndexIdOffset) + ",'" + tenantId + "','" + valuePrefix +
-                        "v2-1'] - ["+(saltBuckets.intValue()-1)+"," + Long.toString(Long.MIN_VALUE + expectedIndexIdOffset) + ",'" + tenantId + "','" + valuePrefix + "v2-1']\n"
+                    "CLIENT PARALLEL 3-WAY RANGE SCAN OVER _IDX_" + tableName + " [0," + Long.toString(Short.MIN_VALUE + expectedIndexIdOffset) + ",'" + tenantId + "','" + valuePrefix +
+                        "v2-1'] - ["+(saltBuckets.intValue()-1)+"," + Long.toString(Short.MIN_VALUE + expectedIndexIdOffset) + ",'" + tenantId + "','" + valuePrefix + "v2-1']\n"
```
Comment (ChinmaySKulkarni) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/BaseTenantSpecificViewIndexIT.java]: ditto here.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/BaseViewIT.java [hunk 6]
```diff
@@ -238,9 +238,9 @@ protected String testUpdatableView(Integer saltBuckets) throws Exception {
         } else {
             physicalTableName = viewIndexPhysicalName;
             assertEquals(saltBuckets == null
-                    ? "CLIENT PARALLEL 1-WAY RANGE SCAN OVER " + viewIndexPhysicalName +" [" + (Long.MIN_VALUE+1) + ",'foo']\n"
+                    ? "CLIENT PARALLEL 1-WAY RANGE SCAN OVER " + viewIndexPhysicalName +" [" + (Short.MIN_VALUE+1) + ",'foo']\n"
                             + "    SERVER FILTER BY FIRST KEY ONLY"
-                            : "CLIENT PARALLEL " + saltBuckets + "-WAY RANGE SCAN OVER " + viewIndexPhysicalName + " [0," + (Long.MIN_VALUE+1) + ",'foo'] - ["+(saltBuckets.intValue()-1)+"," + (Long.MIN_VALUE+1) + ",'foo']\n"
+                            : "CLIENT PARALLEL " + saltBuckets + "-WAY RANGE SCAN OVER " + viewIndexPhysicalName + " [0," + (Short.MIN_VALUE+1) + ",'foo'] - ["+(saltBuckets.intValue()-1)+"," + (Long.MIN_VALUE+1) + ",'foo']\n"
```
Comment (ChinmaySKulkarni) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/BaseViewIT.java]: Shouldn't the last `Long.MIN_VALUE` also be changed to `Short.MIN_VALUE`?

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/UpgradeIT.java [hunk 7]
```diff
@@ -188,7 +188,7 @@ public void testMapTableToNamespaceDuringUpgrade()
             PName physicalName = PNameFactory.newName(hbaseTableName);
             String newSchemaName = MetaDataUtil.getViewIndexSequenceSchemaName(physicalName, true);
             String newSequenceName = MetaDataUtil.getViewIndexSequenceName(physicalName, tenantId, true);
-            verifySequenceValue(null, newSequenceName, newSchemaName, -9223372036854775805L);
+            verifySequenceValue(null, newSequenceName, newSchemaName, -32765);
```
[the remark this answers is not in the document]
Comment (ChinmaySKulkarni) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/UpgradeIT.java]: +1

File: phoenix-core/src/main/java/org/apache/phoenix/coprocessor/BaseScannerRegionObserver.java (hunk 1 above)
Comment (ChinmaySKulkarni) [reply phoenix-core/src/main/java/org/apache/phoenix/coprocessor/BaseScannerRegionObserver.java]: Can you please add a comment here with a short version of the explanation above?

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/index/ShortViewIndexIdIT.java [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+            // we're expecting exactly one row mapping the column value to the key
+            assertTrue(rs.next());
+            assertEquals(3, rs.getInt(1));
+            assertEquals(1, rs.getInt(2));
+            assertFalse(rs.next());
+            rs.close();
+        }
+    }
+
+}
```
Comment (ChinmaySKulkarni) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/index/ShortViewIndexIdIT.java]: Any way we can add a test for the stale region boundaries change as well?

File: phoenix-core/src/main/java/org/apache/phoenix/coprocessor/BaseScannerRegionObserver.java [hunk 9]
```diff
@@ -178,8 +178,11 @@ private static void throwIfScanOutOfRegion(Scan scan, Region region) throws DoNo
...
+            byte[] actualStartRow = scan.getAttribute(SCAN_ACTUAL_START_ROW);
+            isStaleRegionBoundaries = (expectedUpperRegionKey != null &&
+                    Bytes.compareTo(upperExclusiveRegionKey, expectedUpperRegionKey) != 0) || 
+                    (actualStartRow != null && Bytes.compareTo(actualStartRow, lowerInclusiveRegionKey) < 0);
         } else {
             isStaleRegionBoundaries = Bytes.compareTo(lowerInclusiveScanKey, lowerInclusiveRegionKey) < 0 ||
```
Comment (ChinmaySKulkarni) [inline phoenix-core/src/main/java/org/apache/phoenix/coprocessor/BaseScannerRegionObserver.java]: Why is this logic different for the non-local index case? I guess the only thing that should change is how we extract the start and end keys of the scan, right? For ex, in the case of a local index scan, these would be passed as scan attributes and for normal scans they are just the startRow and endRow of the scan, or am I not understanding this correctly?

File: phoenix-core/src/main/java/org/apache/phoenix/iterate/ExplainTable.java [hunk 10]
```diff
@@ -215,7 +215,7 @@ private void appendPKColumnValue(StringBuilder buf, byte[] range, Boolean isNull
     private Long getViewIndexValue(PDataType type, byte[] range) {
         boolean useLongViewIndex = MetaDataUtil.getViewIndexIdDataType().equals(type);
         Object s = type.toObject(range);
-        return (useLongViewIndex ? (Long) s : (Short) s) + (useLongViewIndex ? Long.MAX_VALUE : Short.MAX_VALUE) + 2;
+        return (useLongViewIndex ? (Long) s : (Short) s) + Short.MAX_VALUE + 2;
```
Comment (ChinmaySKulkarni) [inline phoenix-core/src/main/java/org/apache/phoenix/iterate/ExplainTable.java]: I guess this isn't relevant to your change, but is the `+2` for the 'L#' characters? If yes, can we change this to `LOCAL_INDEX_COLUMN_FAMILY_PREFIX.length` for better readability?

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/BaseViewIT.java (hunk 6 above)
Comment (lhofhansl) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/BaseViewIT.java]: Hmm... Good point. Not sure how this test passes.

File: phoenix-core/src/main/java/org/apache/phoenix/coprocessor/BaseScannerRegionObserver.java (hunk 9 above)
Comment (lhofhansl) [reply phoenix-core/src/main/java/org/apache/phoenix/coprocessor/BaseScannerRegionObserver.java]: This is because for local indexes we cannot (ever) proceed when the region was split, we have to restart, since we cannot exactly know where the index keys were split to, they sort differently than the main table's keys.
It's not enough that the new scan is fully "contained" in the new region. Is that end "old" scan boundaries are contained on the lower side and exact on the upper side.

File: phoenix-core/src/main/java/org/apache/phoenix/iterate/ExplainTable.java (hunk 10 above)
Comment (lhofhansl) [reply phoenix-core/src/main/java/org/apache/phoenix/iterate/ExplainTable.java]: I actually don't know. :( I just made sure it behaves like before.

File: phoenix-core/src/main/java/org/apache/phoenix/coprocessor/BaseScannerRegionObserver.java (hunk 1 above)
Comment (ChinmaySKulkarni) [reply phoenix-core/src/main/java/org/apache/phoenix/coprocessor/BaseScannerRegionObserver.java]: @lhofhansl I think that makes sense given what's happening when you get a [StaleRegionBoundaryCacheException](https://github.com/apache/phoenix/blob/f8cfa6b8b9f63ab080cdafa486bb4aabf043b1f8/phoenix-core/src/main/java/org/apache/phoenix/iterate/BaseResultIterators.java#L1315) and call BaseResultIterators#[recreateIterators](https://github.com/apache/phoenix/blob/f8cfa6b8b9f63ab080cdafa486bb4aabf043b1f8/phoenix-core/src/main/java/org/apache/phoenix/iterate/BaseResultIterators.java#L1405)

File: phoenix-core/src/main/java/org/apache/phoenix/util/MetaDataUtil.java (hunk 3 above)
Comment (lhofhansl) [reply phoenix-core/src/main/java/org/apache/phoenix/util/MetaDataUtil.java]: Done.

File: phoenix-core/src/main/java/org/apache/phoenix/coprocessor/BaseScannerRegionObserver.java (hunk 1 above)
Comment (lhofhansl) [reply phoenix-core/src/main/java/org/apache/phoenix/coprocessor/BaseScannerRegionObserver.java]: Yeah. Good idea.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/index/ShortViewIndexIdIT.java (hunk 8 above)
Comment (lhofhansl) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/index/ShortViewIndexIdIT.java]: That's tested in MutableIndexSplitForwardScanIT and MutableIndexSplitReverseScanIT. Those tests fails with the first version of the change.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/BaseTenantSpecificViewIndexIT.java (hunk 4 above)
Comment (lhofhansl) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/BaseTenantSpecificViewIndexIT.java]: I mostly restore this part from before PHOENIX-3547.
Realizing now it was all over the place. Using  + Short.MIN_VALUE is some place, -32768 in others, toString in yet others.
Here it does not matter though, the output is identical.

Comment (lhofhansl) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/BaseTenantSpecificViewIndexIT.java]: In fact this is best expressed as "string" + (number1 + number2) + "string", anyway. I'll make that change.

Comment (ChinmaySKulkarni) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/BaseTenantSpecificViewIndexIT.java]: Sounds good, thanks

File: phoenix-core/src/main/java/org/apache/phoenix/coprocessor/BaseScannerRegionObserver.java (hunk 9 above)
Comment (ChinmaySKulkarni) [reply phoenix-core/src/main/java/org/apache/phoenix/coprocessor/BaseScannerRegionObserver.java]: Makes sense. Thanks for the explanation.

**Comment by ChinmaySKulkarni** [comment]
Closing since JIRA is resolved.

**Comment by stoty** [comment]
Already merged.
